### PR TITLE
Validate tljh specific config

### PR DESCRIPTION
**Pull request by jrdnbradford**
This is a stab at the `FIXME` in `tljh/config.py` for adding `config.yaml` validation. It uses `jsonschema` to validate the changes against it.

This  is a work in progress; I've tried to use the [tljh-config docs](https://tljh.jupyter.org/en/latest/topic/tljh-config.html) and `configurer.py` to fill it out the best I can. 

Example of some current validation:
```bash
sudo tljh-config set user_environment.default_app jupyterlb
'jupyterlb' is not one of ['jupyterlab', 'classic']
```

Certainly not perfect, looking for input before I go any further! 🚀 Also trying to get these tests to pass.

### CODE REVIEW

**Comment by jrdnbradford** [comment]
Looks like I've got most tests passing now.

**Comment by consideRatio** [comment]
Excellent work @jrdnbradford!

How should this behave if existing TLJH installations upgrade to get this validation? They could have invalid config already in place, and then making an unrelated config change may trigger a validation failure.

I'm not confident on what I think makes sense to do here. If we could, I'd like to error loudly and early on before we have actually upgraded anything. I'm not sure we can do that while making use of this validation logic inside the tljh python package, as then we have already upgraded to use it I figure.

At the same time, maybe not. The config could have been manually edited anyhow, and we should assume someone may have done that also...

**Comment by minrk** [comment]
One option to trial this is to have a flag to turn validation warnings into errors or vice versa (e.g. `--no-validate` for opt-out, or `--strict` for opt-in).

**Comment by jrdnbradford** [comment]
@minrk is the suggestion something like this:

- add `--no-validate` and `--strict` flags for `tljh-config` CLI command
  -  `--no-validate` will be the default and will allow edits that don't validate against the schema, but will `print` a warning
  - `--strict` will not allow edits that fail to validate and will `exit` with an error

Those with invalid configs who update TLJH will start to get warnings when they use `tljh-config` but won't get any breaking behavior.

If so, this is currently the best solution I can think of. If there are no objections, I'll add and test the flags.

**Comment by minrk** [comment]
`--[no-]validate` sounds like a good plan (yes, the names should definitely match if we give flags for both ways), thanks @jrdnbradford!

I also think it's just fine to validate by default if that's what others would prefer. I wouldn't call this breaking, because the validation error message should include a message showing how to re-run without validation, along the lines of:

> You can still apply this change without validation by re-running with `--no-validate` (whole copy-pasteable command-line if it's easy).
> if you think this validation error is incorrect, please report it to https://github.com/jupyterhub/the-littlest-jupyterhub/issues

**Comment by consideRatio** [comment]
I agree on the idea to validate by default, so having a flag to opt out of the validation is my preference.

Flag naming question
- `--no-validate` 
- `--no-validation` 
- `--skip-validation`
- `--skip-config-validation`

I'm leaning for the most verbose flag of `--skip-config-validation` which clarifies that its about config validation specifically - because one could imagine other kinds of validation be made as well otherwise.

**Comment by minrk** [comment]
I like the verb forms, and I think it's best to stick the very standard `no-` prefix, and not uncommon `skip-`. So I'd go with `--[no-]validate`, or if you think it's really likely that we'll have more than one optional thing to validate, `--[no-]validate-config`.

**Comment by jrdnbradford** [comment]
Got an initial working version using `--[no-]validate`. The one failed integration test is because Python `v3.8` doesn't have the `argparse.BooleanOptionalAction` option. 

I also bumped the `Dockerfile` image to `22.04`.

**Comment by jrdnbradford** [comment]
@consideRatio and @minrk if the failed Python 3.8 test is a blocker, I can rewrite to make it compatible. Let me know!

**Comment by minrk** [comment]
Yes, that would be great if you can make it work on 3.8. Thank you!

**Comment by jrdnbradford** [comment]
@consideRatio and @minrk the Python 3.8 test is now passing after removing `argparse.BooleanOptionalAction`. I also fixed the two unit tests that were failing (they were just missing `pip` so they couldn't install dependencies). Not sure if the `Integration tests / Ubuntu 22.04, Py 3.10, from 0.2.0 (pull_request) ` test is still needed.

Let me know if I can do anything else here. 🚀

File: tljh/config.py [hunk 1]
```diff
@@ -380,16 +412,25 @@ def main(argv=None):
 
     args = argparser.parse_args(argv)
 
+    if args.validate == None:
+        args.validate = True
```
[the remark this answers is not in the document]
Comment (jrdnbradford) [reply tljh/config.py]: It is not! Pushed a removal.

File: integration-tests/test_proxy.py [hunk 2]
```diff
@@ -63,9 +63,9 @@ def test_manual_https(preserve_config):
             "/CN=tljh.jupyer.org",
         ]
     )
-    set_config_value(CONFIG_FILE, "https.enabled", True)
-    set_config_value(CONFIG_FILE, "https.tls.key", key)
-    set_config_value(CONFIG_FILE, "https.tls.cert", cert)
+    set_config_value(CONFIG_FILE, "https.enabled", True, True)
+    set_config_value(CONFIG_FILE, "https.tls.key", key, True)
+    set_config_value(CONFIG_FILE, "https.tls.cert", cert, True)
```
[the remark this answers is not in the document]
Comment (jrdnbradford) [reply integration-tests/test_proxy.py]: Will do.

**Comment by consideRatio** [comment]
@jrdnbradford a trick to combine multiple code suggestions is to visit "files changed", from there you can "add suggestion to batch" and add multiple code suggestions in one go.

I'm actively debugging the remaining test failures now btw

**Comment by jrdnbradford** [comment]
Refactoring done. Good to merge when you're ready. Thanks, everyone!

**Review by consideRatio — APPROVED** [review]
THANK YOU @jrdnbradford!!!

I find robustifying contributions like these critical for the health of a open source projects - thank you :heart: :tada: :sunflower:!!

**Pull request merged**